CircleCI windows builds

venimus opened the pull request:
### Changes:
* Migrate from Appveyor to CircleCI
* New install and build scripts (usable locally)
* Removed Java requirements
* Docs for windows builds are updated

### Detailed notes:

Migrate from Appveyor to CircleCI:
- Workflow for windows builds with OTP21.3. Version is configurable. In general it should not differ from Appveyor.
- Cache (layered) of deps folders. Force cache rebuild by bumping `&windows_cache_version`
- Newest windows orbs already contains VS2019.
- Unfortunately CI jobs fail (with a timeout error) when a background process is left from an installer or build. So some stuff was refactored to avoid such leftovers (i.e. manually download and control installers and test script fixes).
- Tagging of new Aeternity versions no longer require to change package.cfg. It is now automatically resolved.
- Windows jobs are controlled by parameters, so they can be reused without much changes to the config. Adding different ENV builds (i.e. different Erlang version) is just by adding another set of aliases for `*windows_build` and `*windows_test` jobs and use them in `jobs:` with the proper parameters. Unfortunately using aliases is the only way (currently) to reuse jobs.
- CI build scripts are now under .circleci/windows (ci/appveyor are kept for BC but will be removed eventually). They rely on the scripts designed for users.

Building scripts:
- Scripts are refactored and fixed to be more user-friendly and keep CI-friendly at the same time. This will increase consistency among build environments. Thus CI `build.bat` script could be used locally to build the artifacts (although it is not meant for that).
- ENV preparation script (`msys2_prepare.bat`) is refactored and split to several standalone scripts that could be used by users (or CI)
- MSYS scripts could be used to tune and handle different installation folders much easier (and faster). It also have help (--help) and debug (-v) mode used by CI.
- Erlang/OTP script now downloads from a much faster mirror (by default). Downloading from the official site is incredibly slow. Script now allows installation of different versions (multiple as well) by providing params or ENV vars. So it is suitable for both users or CI.
- Certain ENV variables could be used to control the build, most of them are optional and used for speeding up the builds: i.e. some of them represent different paths and if undefined, the scripts will try to resolve automatically, which takes some time. Those are set in CircleCI config.
- Docs for windows builds are improved. Although nothing much changed from the user point of view.

Too many changes were needed for this migration, because old scripts were not flexible enough to be fixed to work with new env. That is why I decided to improve the user tools and base the CI builds on them. In fact now CI would mostly mimic how a user could build a local release.
*Note: Windows shell scripting is quite messy and certain basic tasks need some hacks.*
*Note: msys2_prepare seems to be copied from erlang docs, but originally it is designed for building erlang itself. So there is no reason to keep it in sync with it, furthermore the origin is badly written anyway. However I tried to not change it much and I think there are no any breaking changes in this PR.*
Originally branched v4.2.0 and then rebased to master (both work ok)

venimus commented on ci/windows/package.cfg:
```diff
@@ -0,0 +1,79 @@
+
+description = The full Aeternity node for Windows.
+
+version = 4.2.0
```
This should be changed when new version is introduced

venimus replied:
This is now automatically resolved

venimus commented on .circleci/windows/erlang_install.bat:
```diff
@@ -0,0 +1,20 @@
+@echo %time% : Install Erlang/OTP %1
+IF EXIST %2 GOTO:INSTALLED
+
+SET "URL=https://packages.erlang-solutions.com/erlang/erlang/esl-erlang/FLAVOUR_1_general/esl-erlang_%1~windows_amd64.exe"
```
This is a much faster download mirror, not sure if reliable. Official site download takes some minutes vs few seconds here

venimus commented on .circleci/windows/erlang_install.bat:
```diff
@@ -0,0 +1,20 @@
+IF NOT "%ERRORLEVEL%"=="0" (echo %time% : Download failed && exit /b %ERRORLEVEL%)
+
+@echo %time% : Install started
+powershell start-process -wait '%temp%\erl%1.exe' -argumentlist '/S /D=%2'
```
since the installer always runs as a separate process we need to hold until it finish

venimus commented on .circleci/windows/erlang_install.bat:
```diff
@@ -0,0 +1,20 @@
+@echo off
+IF NOT "x%1"=="x" IF NOT "x%2"=="x" GOTO:INSTALL
+@echo Usage: %0 ^<otp_version^> ^<install_path^>
```
although this is CI script i made it user friendly so it could be reused locally by users

venimus commented on .circleci/windows/msys2_install.bat:
```diff
@@ -0,0 +1,35 @@
+md "%WIN_MSYS2_ROOT%"
+pushd "%WIN_MSYS2_ROOT%\.."
+
+curl -s -O https://tukaani.org/xz/xz-5.2.4-windows.zip
```
MSYS2 is distributed as this inadequate (for windows) .xz format which requires an additional tool to decompress it.
I didn't want to use 7z because it might not be available

venimus commented on .circleci/windows/msys2_install.bat:
```diff
@@ -0,0 +1,35 @@
+pushd "%WIN_MSYS2_ROOT%\.."
+
+curl -s -O https://tukaani.org/xz/xz-5.2.4-windows.zip
+tar -xf xz-5.2.4-windows.zip
```
tar is available in Windows 10

venimus commented on .circleci/windows/msys2_install.bat:
```diff
@@ -0,0 +1,35 @@
+popd
+
+@echo %time%: Msys2 Initial setup
+bash -lc "echo Done."
```
First run does some preparations of MSYS so I just dump tracking text

venimus commented on .circleci/windows/msys2_install.bat:
```diff
@@ -0,0 +1,35 @@
+bash -lc "echo Done."
+
+@echo %time%: Msys2 Update 1/2
+bash -lc "pacman --noconfirm -Syuu"
```
msys requires few runs of pacman to complete the installation.

venimus commented on .circleci/windows/msys2_prepare.bat:
```diff
@@ -0,0 +1,4 @@
+@echo on
+git config --global --unset url."ssh://git@github.com".insteadOf "https://github.com"
```
CircleCI by default has set this in git config, but it prevents to download from public repos (specifically styrene) without a key

venimus commented on .circleci/windows/package.cfg:
```diff
@@ -0,0 +1,79 @@
+
+description = The full Aeternity node for Windows.
+
+version = VERSION_PLACEHOLDER
```
This will be replaced by the node version automatically, to avoid changing the script with every release

venimus commented on .circleci/windows/test.bat:
```diff
@@ -0,0 +1,66 @@
+@echo on
```
This script simulates user behavior. It extracts the .zip artifact and runs smoke tests

venimus commented on .circleci/windows/test.bat:
```diff
@@ -0,0 +1,66 @@
+@echo %time% : Run test: release
+bash -lc ^"cd ${BUILD_PATH} ^&^& ^
+           make python-env PIP=/mingw64/bin/pip3 ^&^& ^
+           rm -rf ${PACKAGE_TESTS_DIR} ^&^& ^
```
I did remove `epmd -daemon` as it make CircleCI wait forever after the job finishes (I reported them this, but don't expect much) I guess CircleCI finds that there is a background process that is executing and will wait for it but will fail after 10 minutes with "No output for 10 minutes" message.
Unfortunately using 'epmd -kill' didn't fix the problem.

venimus commented on .circleci/windows/vcvarsall.bat:
```diff
@@ -0,0 +1,19 @@
+@echo on
```
this helper script finds and executes `vcvarsall.bat` which is located on different places depending on the VS version and sets MSCV_VERSION var to the proper value.

venimus commented on scripts/windows/msys2_shell.bat:
```diff
@@ -23,21 +23,23 @@ FOR /F "tokens=* USEBACKQ delims=" %%F IN (`type "%ToolsVerFile%"`) DO SET MSVC_
 rem Open shell
 
-%WIN_MSYS2_ROOT%\mingw64.exe
+"%WIN_MSYS2_ROOT%\mingw64.exe"
```
path might include spaces and it will fail without `"`

venimus commented on .circleci/config.yml:
```diff
@@ -434,6 +476,111 @@ commands:
+          erts_version: << parameters.erts_version >>
+          install_path: << parameters.install_path >>
+
+  windows_install_deps:
```
OTP and MSYS are split to separate steps, so they can use different caches (this saves some build time) and allows different jobs with different Erlang versions to be cached properly

venimus commented on .circleci/config.yml:
```diff
@@ -960,13 +1107,108 @@ jobs:
+          path: *windows_artifacts
+          destination: /tmp/packages-<< parameters.otp_version >>
+
+  win_build_20_1: *windows_build
```
We need to create something like aliases for the jobs, because config requires every job to have unique key. This is kind of hackish

venimus commented on .circleci/windows/build.bat:
```diff
@@ -0,0 +1,34 @@
+exit /b %ERRORLEVEL%
+
+:log :: Display a log message
+@echo :: [1;33m %time% : %* [0m>con && exit /b
```
Unfortunately there is no way to send escaped ESC sequences in windows, so we need to put the ESC char as is. It is a valid ASCII char anyway. This adds some color in the build log and makes it much easier to inspect/debug

venimus commented on .circleci/windows/build.sh:
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
```
This script was extracted from build.bat (which was simply running bash commands)

venimus commented on .circleci/windows/build.sh:
```diff
@@ -0,0 +1,64 @@
+
+echo -e "${bold}Create package environment${reset}"
+NODE_VERSION=`cat ${PROJECT_ROOT}/VERSION`
+sed "s/^version =.*$/version = ${NODE_VERSION}/g" .circleci/windows/package.cfg > ${WORK_ROOT}/package.cfg
```
Resolve node's version and put it in package.cfg

venimus commented on .circleci/windows/build.sh:
```diff
@@ -0,0 +1,64 @@
+
+# Copy genesis and hard-fork account migrations into top-level data folder
+mkdir -p ${WORK_PATH}/data
+cp -R ${RELEASE_PATH}/data/aecore ${WORK_PATH}/data/
```
I couldn't find why the old script was copying the genesis accounts implicitly one by one.

venimus commented on .circleci/windows/msys2_prepare.bat:
```diff
@@ -0,0 +1,6 @@
+@echo on
+:: Unset CircleCI default setting that prevents pulling Styrene from the public repo
+git config --global --unset url."ssh://git@github.com".insteadOf "https://github.com"
```
CircleCI sets `git config --global url.ssh://git@github.com.insteadOf https://github.com` on the 'checkout' step and forgets to remove it. However it prevents to download from public repos without a key. So we need to revert this setting to be able to checkout Styrene (which has no release binary)

venimus commented on .circleci/windows/package.cfg:
```diff
@@ -0,0 +1,79 @@
+
+description = The full Aeternity node for Windows.
+
+version = VERSION_PLACEHOLDER
```
Version is no longer hard-coded and it will be resolved on build time

dincho replied:
I don't really like the `VERSION_PLACEHOLDER`, perhaps I'd expect `master` or something "real" - but OK

venimus replied:
it does not really matter what it will be anyway. It does not stop building the package. Also replace will look for `version =`. While missing this step and using it directly it will make this more obvious for debugging

venimus commented on .circleci/windows/test.bat:
```diff
@@ -0,0 +1,70 @@
+exit /b %ERRORLEVEL%
+
+:: Subroutines
+:TEST_ct
```
These steps are not used by CI but I kept them for BC

venimus commented on .circleci/windows/test.bat:
```diff
@@ -0,0 +1,70 @@
+SET "PATH=%WIN_MSYS2_ROOT%\mingw64\bin;%WIN_MSYS2_ROOT%\usr\bin;%WIN_MSYS2_ROOT%;%PATH%"
+SET "PATH=%WIN_OTP_PATH%\bin;%WIN_OTP_PATH%\erts-%ERTS_VERSION%\bin;%WIN_OTP_PATH%\erts-%ERTS_VERSION%;%PATH%"
+
+:: Construct unix paths
```
Next set of commands will try to resolve where the artifacts are, so this script could be used locally. In addition because msys is a posix environment we need to prepare both versions of windows and unix forms of the paths

venimus commented on scripts/windows/install_erlang.bat:
```diff
@@ -0,0 +1,73 @@
+PowerShell (New-Object System.Net.WebClient).DownloadFile('%OTP_URL%', '%TMP%\erl.exe')
+@IF NOT "%ERRORLEVEL%"=="0" (call:log Download failed && exit /b %ERRORLEVEL%)
+
+:: Uninstall the old version
```
Just in case we have old version of Erlang in the same directory we have to uninstall it first. If we install in a separate folder <install_path> we can have multiple versions simultaneously. Then switching the active can be done via ENV

venimus commented on scripts/windows/install_msys2.bat:
```diff
@@ -0,0 +1,66 @@
+@call:log Target directory %WIN_MSYS2_ROOT%
+
+IF EXIST "%WIN_MSYS2_ROOT%" rd /s/q "%WIN_MSYS2_ROOT%"
+SET "PSAdmin=powershell Start-Process -wait -WindowStyle hidden 'powershell.exe'"
```
This 'Inception' command is to be able to prepare a self-elevating shell for executing stuff requiring admin rights

venimus commented on scripts/windows/install_msys2.bat:
```diff
@@ -0,0 +1,66 @@
+powershell (New-Object System.Net.WebClient).DownloadFile('%MSYS2_URL%', '%TMP%\msys2.tar.xz')
+
+@call:log Download PowerShell 7z helper
+%PSAdmin% 'Install-Package -Scope CurrentUser -Force 7Zip4PowerShell'
```
Unfortunately msys2 is distributed in this inadequate .xz compressed format and we need some way to extract it. Seems the easiest way is to use this PS module that adds 7z to powershell

venimus commented on scripts/windows/install_msys2.bat:
```diff
@@ -0,0 +1,66 @@
+@echo off
+:: Pass -v as first argument to enable debug echo
+:: Script to download and install Msys2 from %MSYS2_URL%
```
Installing MSYS2 is quite complex and does not have an installer, so we need our own script to do it. It could be very helpful for users too

venimus commented on scripts/windows/msys2_prepare.bat:
```diff
@@ -11,26 +13,101 @@
+
+:: Set the paths appropriately avoiding duplicates
+SET "_PATHS=%WIN_MSYS2_ROOT%\mingw64\bin;%WIN_MSYS2_ROOT%\usr\bin;%WIN_MSYS2_ROOT%"
+ECHO."%PATH%" | FINDSTR /V /C:"%_PATHS%">nul && SET "PATH=%_PATHS%;%PATH%"
```
This "ugly" mess is to check if we already have changed the PATH in a previous run. Users may run this script multiple times and we don't want to fill the PATH with duplicates. Shockingly it has a limit of 1024 chars. Visual studio preparation script that is executed earlier shamelessly adds dozens of entries already

venimus commented on scripts/windows/msys2_prepare.bat:
```diff
@@ -1,4 +1,6 @@
 @echo off
```
Changes in this script are basically to add more checks for env and refactor it to subroutines for easier maintenance and clarity. It does not do anything fundamentally different from the old one

venimus commented on scripts/windows/msys2_prepare.bat:
```diff
@@ -66,74 +143,41 @@ mingw-w64-x86_64-python3-pip ^
 
-echo Current time: %time%
-rem Set up msys2 env variables
+SET BASH="%WIN_MSYS2_ROOT%\usr\bin\bash.exe"
```
Unfortunately setting Path is not always enough to guarantee that we execute the proper bash (i.e. we can have WSL, or git bash already installed). So we better implicitly refer to the executable.

venimus commented:
I'm not sure if it is not better to build with otp 21.0 instead of 20.1

venimus commented:
Latest commits:
* Rebased and added OTP21.3 build (also removed redundant ci params)
* In addition added some UX improvements to install scripts
* Updated docs

venimus commented on scripts/windows/install_erlang.bat:
```diff
@@ -0,0 +1,82 @@
+@echo off
```
This script is to streamline installation of erlang supporting multiple simultaneous versions and uses erlang-solutions as mirror (by default) as it is way too faster to download. CI download and install will require a script anyway. Chocolatey script does not support installing multiple versions properly and also leaves a running process which bugs the job execution (waits forever for it to finish)

venimus commented on scripts/windows/msys2_prepare.bat:
```diff
@@ -11,26 +12,119 @@
+IF "%WIN_MSYS2_ROOT%"=="" SET "WIN_MSYS2_ROOT=C:\tools\msys64"
+
+IF "%FORCE_STYRENE_REINSTALL%"=="" SET "FORCE_STYRENE_REINSTALL=false"
+IF "%JDK_URL%"=="" SET "JDK_URL=https://download.java.net/java/GA/jdk11/9/GPL/openjdk-11.0.2_windows-x64_bin.zip"
```
It is probably not used in CI, but this script is also used by 'users/devs' which may do other tasks that require it -> build UML diagrams, so we need to get rid of it there first. Or just let them install it manually as with Visual Studio

venimus commented on .circleci/windows/package.cfg:
```diff
@@ -0,0 +1,79 @@
+[rename_db.desktop]
+name = rename_db
+comment = rename_db utility for Aeternity
+exec = ./usr/lib/aeternity/bin/aeternity.cmd rename_db
```
I'm not sure which of these are still valid (and what they should do). May be someone more qualified should check it. @tolbrino ?

venimus commented:
> I see 3 windows build jobs using 3 different OTP versions - it does not make sense to me:

Yep, It is the main concern to me too. I actually left this just to decide which one should stay, as only one of the resulting packages will be distributed anyway.
I did split the test to make sure we use clean environment. But if the CPU-time is an issue there is no problem to merge it.
There is also a possibility to have 1 job that do 2-3 builds at the same time

venimus commented:
> Blocking mainly because of the old `rename_db` command and multiple build jobs.

I merged both steps and only left the 20.3 build as it seems most stablehttps://github.com/aeternity/aeternity/pull/2692/commits/3076b9298c25bf42581c2eb0f8a29bbe859449fa.

And removed the cmd https://github.com/aeternity/aeternity/pull/2692/commits/aac271eed91b01d4b20cbff9a538e950b4b66e0d

venimus commented on .circleci/config.yml:
```diff
@@ -952,13 +1090,41 @@ jobs:
   build_test_deploy:
     jobs:
+      - windows_package:
+          otp_version: "20.3"
```
I would too, but was not sure if it is officially supported

venimus commented on .circleci/config.yml:
```diff
@@ -399,6 +403,40 @@ references:
+    OTP_VERSION: << parameters.otp_version >>
+    WIN_OTP_PATH: << parameters.otp_install_path >>
+    WIN_MSYS2_ROOT: << parameters.msys_install_path >>
+    JAVA_VERSION: 12.0.2
```
I left it so preparation script will skip installing it. It is already preinstalled in the orb.
May be I will remove it from there and note to devs that they should install it manually if needed

venimus commented:
last commit:
* Removed Java requirements (and references) and updated doc
* Renamed most .bat files to .cmd (as they have cmd extensions by default)
I didn't rename `msys2_prepare.bat` and `msys2_shell.bat` for BC
* Switched build to OTP 21.3
